man/dbmacro.1
.TH DBMACRO 1 "20 January 2000"
.FB
.de SF
.nf
.nj
.sp 1
.ft CW
.in +3
..
.de UF
.XF
.sp 1
..
.de XF
.ft
.ju
.fi
.in -3
..
.SH NAME
dbmacro \- general purpose text, CGI, HTML, and record processing language
.SH SYNOPSIS
.B dbmacro
[\fB-b\fP]
[\fB-c\fP]
[\fB-r\fP]
[\fB-d\fP \fIdatabase\fP]
[\fB-i\fP \fIindex\fP]
[\fImacro_file\fP | \fB-m\fP \fIscript\fP]
.br
.SH DESCRIPTION
\fIDbmacro\fP, a general purpose processing tool for
text files, CGI requests, HTML forms, database records and
\fIFirstBase\fP databases.
\fIDbmacro\fP executes \fImacro\fP(5) code from a file, standard input,
or an inline script. Default output is written to standard output.
.PP
The \fImacro\fP
language provides a powerful general purpose processing tool that
can be used with or without a FirstBase database.
.PP
The \fImacro\fP(5) language is similar to C in syntax
and includes most of the familiar programming constructs,
including conditionals, loop constructs, user defined functions,
and printf mechanisms.
This high level language also supports
associative arrays, run time loadable code, strings, and
all of the \fIFirstBase\fP data types. In addition, there are
many built in functions covering math, string, date, screen input/output,
stream input/output, and CGI/HTML processing.
.PP
Variables are dynamically created without being declared and
are global in all situations unless specified as local
using the \fIlocal\fP directive. These variables can take on any of
these data types: float, date, string, associative array, or database array.
.PP
Functions can accept and return
any number of arguments and parameters of any type, including arrays.
All function parameters are local.
.PP
At runtime, variables can be introduced and initialized
on the argument line using a \fIvar=value\fP syntax. Variables can also be
introduced using CGI methods. See below.
.PP
Database records and fields can be created, modified, or deleted
using simple assignments and/or calls to the \fIputrec\fP function.
The \fB-r\fP flag invokes \fIdbmacro\fP in read-only mode so that
records cannot be modified.
.PP
Using the \fB-c\fP flag, or
invoking \fIdbmacro\fP as \fIdbmacro.cgi\fP, enables \fIdbmacro\fP to
process Common Gateway Interface (CGI) requests, either GET or POST methods,
as used by http servers and World Wide Web browsers.
See the section on CGI below.
.PP
A \fImacro_file\fP is a simple text file containing \fImacro\fP(5) code.
Generally, a \fImacro_file\fP can contain three main sections:
\fIbegin\fP, \fIbody\fP, and \fIend\fP, and any number
of \fIfunction\fP sections.
Each of these sections is declared using these keywords. The default
section type is \fIbody\fP.
.PP
When provided with a database argument, \fIdbmacro\fP will
processes the database in the normal \fIFirstBase\fP generator manner
\(em the macro_file is executed once for each referenced record.
The set of processed records can be limited using the \fB-i\fP index specifier.
The \fIbegin\fP section is done once before any processing occurs
(for initialization needs), the \fIbody\fP section is done once for each
referenced database record
and the \fIend\fP section is done once after the last record
has been processed (for clean up or totals printing).
.PP
When not provided with a database, \fIdbmacro\fP processes the
\fImacro_file\fP one time.
In this scenario, if database processing is desired, \fImacro\fP(5) functions
like \fIopendb\fP and \fIgetxrec\fP are used in the \fImacro\fP code.
.PP
When using the \fB-m\fP method of providing \fImacro\fP code,
use single quote marks to surround the code.
Note: this method assumes that all code in the script is part of
the \fIbody\fP section.
.PP
\fIDbmacro\fP can be used for database processing,
simple reports, complex inquiries involving many criteria,
and selective, sophisticated database updates.
Fields in records can be modified
by using code that does simple field assignment.
.PP
The following example shows a simple \fImacro\fP file. Note that this BEGIN
section is not really needed (since all variables default to 0 or the NULL
string) but, it serves the purpose here.
.SF
begin
total = records = 0

body
records++
total += $AmountDue

end
print("The grand total is ", formfield(total, "$", 18))
print("The average is ", formfield(total/records, "$", 18))
print("The number of records processed is ", records)
.UF
This example will calculate the grand total of a \fIdollar\fP field named
\fIAmountDue\fP. After processing, the total will be formatted
and printed with a message, along with an average and number of records.
To use this \fImacro\fP
code, the database must be specified using the \fB-d\fP flag.
.PP
As another example, the following is
a functionally equivalent piece of \fImacro\fP(5) code. In this case
\fIdbmacro\fP is invoked without a database,
but uses builtin macro functions to directly open and process a
\fIFirstBase\fP database.
.SF
#
# sample code to read and total records
#
db = opendb("dbase", "r")
if (db >= 0){
   process(db)
   closedb(db)
   }
else
   print("could not open the database")
# end of main section

#
# process - process all records of database db
#
function process(db)
   {
      total = records = 0
      initrec(db, dbase)
      records = reccnt(db)
      for (rec = 1; rec <= records; rec++){
         getrec(rec, db)
         total += dbase[AmountDue]
         }
      print("The grand total is ", formfield(total, "$", 18))
      print("The average is ", formfield(total/records, "$", 18))
      print("The number of records processed is ", records)
   }
.UF
.SH COMMON GATEWAY INTERFACE (CGI)
\fIDbmacro\fP will process both GET and POST requests via the Common
Gateway Interface (CGI). Install \fIdbmacro\fP into the http server's
cgi-bin directory, rename \fIdbmacro\fP to \fIdbmacro.cgi\fP,
and allow access to \fIdbmacro.cgi\fP via the http server.
In addition, create a symbolic link from \fI/usr/local/firstbase\fP to
the actual $FIRSTBASEHOME directory (if they are different).
.PP
If invoked as \fIdbmacro.cgi\fP,
CGI data is automatically converted to variables available
in the macro code at runtime. In this case, \fIdbmacro\fP(1) also
emits the standard CGI script lines "Content-type: text/html\\n\\n".
Multipart CGI data is also handled automatically (see \fIhtml_form_open()\fP
in \fImacro\fP(5)).
.PP
However, if \fIdbmacro\fP(1) is called from a shell script,
or is used as the command interpreter itself via
something like
.SF
#!/usr/local/firstbase/bin/dbmacro
.UF
as the first line of the CGI file, then the \fIcgi_read\fP mechanism
can be used to parse standard CGI parameters and values.
Note that the script will have to print the HTML Content-type directive
itself if needed.
.PP
When processing a GET request, the query string will be searched for
\fIvar=value\fP pairs. During a POST request, all HTML form fields
are used as \fIvar=value\fP pairs. In all cases, 
the variable \fIvar\fP will be preassigned the string \fIvalue\fP
before \fImacro\fP processing begins.
.PP
To pass the name of a file with the \fImacro\fP
code to execute, connect the \fImacro\fP
file name to the URL just beyond the \fIdbmacro.cgi\fP string.
.PP
In the following CGI GET request example,
\fIdbmacro\fP executes the \fImacro\fP code in \fIscript.m\fP
passing a variable named \fICommand\fP that has a string value of
\fIModify\fP, and another variable, \fIStatus\fP, set to \fIOK\fP.
.SF
/cgi-bin/dbmacro.cgi/script.m?Command=Modify&Status=OK
.UF
.PP
In the following CGI POST request example,
\fIdbmacro\fP uses the \fImacro\fP code in the file
\fIwebmaker.m\fP. Variables are passed from standard HTML input fields, in this
case, a field named \fICommand\fP, and another named \fIStatus\fP.
The values of these fields will be whatever is contained in the HTML form
field at the time of the POST request.
.SF
<FORM METHOD="POST"
   ACTION="www.horseweb.com/cgi-bin/dbmacro.cgi/webmaker.m">
<INPUT TYPE=TEXT NAME=Command SIZE=10 MAXLENGTH=10>
<INPUT TYPE=HIDDEN NAME=Status VALUE=OK>
.UF
.PP
The special \fIcheckbox\fP HTML forms input type is provided to
\fIdbmacro\fP code as an array named the same name as the field name.
(The same applies to the use the keyword MULTIPLE in
an HTML SELECT statement).
The subscripts or keys of this checkbox array are from
\fIone\fP to \fIcountkey(array)\fP.
In other words, \fIdbmacro\fP handles multiple occurrences of the same
variable passed in from an HTML form.
.PP
However, there is one caveat with these methods where the same
variable has multiple values as passed in from an HTML form:
If only one of many in the HTML \fIcheckbox\fP or selection list is
selected, then
\fIdbmacro\fP will only see one occurrence of this HTML field name,
and the variable available at run time will not be an array, but will
be a simple variable like any other HTML input fields. In this case, the
\fIcountkey\fP function will return a zero (0).
.PP
If provided with a variable named \fIdatabase\fP, the value of this variable
will be used as the name of the default database, and \fIdbmacro\fP will
behave as if it received a \fB-d\fP flag. Another variable, \fIindex\fP, can
be used to process the default database via index.
.PP
Two other variables are automatically set as well: PATH_TRANSLATED (or
path_translated) is set to the full path name of the excess values beyond
the \fIdbmacro.cgi\fP reference, for example,
\fI/usr/htdocs/firstbase/webmaker.m\fP. In addition, the base directory name
is stored into PATH_DIR (or path_dir).
.PP
File type variables from encoded multipart forms
will contain the name of a temporary file with the contents
of the CGI transmission.
See \fIhtml_form_open()\fP in \fImacro\fP(5) for more details.
.SH MORE EXAMPLES
Here are some more example uses of dbmacro. In the following example,
a database is opened, ten records are added, and the database is closed.
.SF
db = opendb("dbase", "w")
for (i = 1; i < 10; i++){
   initrec(dbase, db)
   dbase[1] = i
   dbase[2] = date(now())
   st = addrec(db)
   printf("status from addrec is %d\\n", st)
   }
closedb(db)
.UF
.PP
In this next example, a database is opened, an index is requested, and
an array named \fIdbase\fP is associated with the database via
\fIinitrec\fP.
Then, a record with a search key of value \fI6\fP is located, printed,
locked, deleted, and then unlocked. Finally, the database is closed.
.SF
db = opendb("dbase", "w")

st = useidx(i, channel)
printf("status of useidx is %d\\n", st)

initrec(dbase, db)

key = makess("6", "N", 6)
st = getxrec(key, db)
if (st < 0){
   printf("could not find record ...exiting\\n")
   exit(1)
   }
printf("%s %s %s %s\\n", dbase[1], dbase[2], dbase[3], dbase[4])

rec = recno(db)
st = lock(rec, db)
printf("st for lock is %d ... pausing - hit return to continue", st)
gets()

st = delrec(db);
printf("st for delrec is %d\\n", st)

unlock(rec, db)

closedb(db)
.UF
.PP
In the next example, no databases are used at all. Here, an array is
subscripted with a couple of strings, and initialized. Then, the array
is processed twice, printing each element.
.SF
word["apple"] = "A good fruit"
word["orange"] = "An orange fruit"

for (;;){
   if (key(subscript, word))
      printf("got a subscript of %s, val of %s\\n", subscript, word[subscript])
   else
      break;
   }
print("Start again")
for (;;){
   if (!key(subscript, word))
      break;
   printf("got a subscript of %s, val of %s\\n", subscript, word[subscript])
   }
.UF
.SH SEE ALSO
dbedit(1), macro(5), fbtstmac(8)
.PP
.I FirstBase User's Guide and Reference Manual
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/dbtools/dbmacro.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: dbmacro.c,v 9.4 2005/01/04 23:31:50 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char Dbmacro_sid[] = "@(#) $Id: dbmacro.c,v 9.4 2005/01/04 23:31:50 john Exp $";
#endif

/*  
 *  dbmacro.c - engine for the dbmacro tool
 */

#include <fb.h>
#include <fb_vars.h>
#include <macro_v.h>
#include <signal.h>
#include <setjmp.h>

extern short int cdb_secure;
extern short int cdb_returnerror;
extern short int cdb_lockmessage;
extern short int cdb_usrlog;
extern short int cdb_limited;
extern short int cdb_locktime;
extern char *cdb_logfile;

static short int cdb_locktime_cgi = 300;

static char *USAGE =
   "usage: dbmacro [-d dbase] [-i index] [-c] [-r] [macrofile | -m script]";
static long n;
short int cdb_write_it = 0;
short int read_only = 0;
short int trace_flag = 0;
int mode;

extern short int cdb_cgi_flag;

#if !FB_PROTOTYPES
static macro();
static initmacro();
static process();
static s_macro();
static writerec();
static void usage();
#else
int main(int, char **);
static int macro(char *, char *);
static int initmacro(char *, char *);
static int process(fb_database *);
static int s_macro(fb_database *);
static int writerec(fb_database *);
static void usage(void);
#endif

static short int withdatabase = 0, withindex = 0;
static fb_mnode *s_exec;
static jmp_buf jmp_env;
static RETSIGTYPE dbmacro_sigalrm(int disp);

/*
 *  dbmacro - main driver
 */
 
   main(argc, argv)
      int argc;
      char *argv[];
   
      {
	 int i;
         char fname[FB_MAXNAME], iname[FB_MAXLINE], *p, var[FB_MAXLINE];
         char word[FB_MAXLINE], *mscript = NULL, path_trans[FB_MAXLINE];
         char path_dir[FB_MAXLINE];

         (void) Dbmacro_sid;

         cdb_batchmode = 1;
         cdb_returnerror = 1;
         fname[0] = NULL;
         iname[0] = NULL;
         fb_basename(fname, argv[0]);
         i = strlen(fname);
         if (equal(fname+i-4, ".cgi"))
            cdb_cgi_flag = 1;
         fname[0] = NULL;
         cdb_db = (fb_database *) NULL;
         p_symtab = symtab = fb_makesymtab();
         g_symtab = fb_makesymtab();
         for (i = 1; i < argc; i++){
            if (argv[i][0] == CHAR_MINUS){
               if (argv[i][2] == NULL){
                  switch (argv[i][1]) {
                     case 'c':
                        cdb_cgi_flag = 1;
                        break;
                     case 'r':
                        read_only = 1;
                        break;
                     case 'd':
                        withdatabase = 1;
                        i++;
                        break;
                     case 'i':
                        withindex = 1;
                        i++;
                        break;
                     case 'x':
                        trace_flag = 1;
                        break;
                     case 'm':
                        i++;
                        mscript = argv[i];
                        break;
                     }
                  }
               }
	    else if ((p = strchr(argv[i], '=')) != 0){
	       *p = NULL;
	       strcpy(var, argv[i]);
	       strcpy(word, p + 1);
	       mf_set_value(var, word);
	       *p = '=';
               }
            else
               strcpy(fname, argv[i]);
            }

         if (cdb_cgi_flag){
            fb_getargs(argc, argv, FB_NODB);
            
            /*
             * this can be used to try and debug why
             * some cgi requests appear to hang.
             *   fb_setup_argv(argc, argv);
             *   fb_writeable_homefile(&cdb_logfile, "USRLOG");
             *   if (cdb_usrlog > 5)
             *      fb_usrlog_begin(argc, argv);
             */

            fb_cgi_init();

            if (fb_cgi_read() != FB_AOK){
               printf("<P>Error reading cgi input\n");
               fflush(stdout);
               exit(-1);
               }
               
            /*
             * use this to test if needed.
             * fb_cgi_echo(); exit();
             */

            /* PATH_TRANSLATED is the cgi fname */
            path_dir[0] = NULL;
            path_trans[0] = NULL;
            p = getenv("PATH_TRANSLATED");
            if (p){
               strcpy(path_trans, p);
               fb_dirname(path_dir, path_trans);
               mf_set_value("PATH_TRANSLATED", path_trans);
               mf_set_value("path_translated", path_trans);
               mf_set_value("PATH_DIR", path_dir);
               mf_set_value("path_dir", path_dir);
               }

            /* fb_getargs(argc, argv, FB_NODB); */

            /* get the database and index field */
            fb_cgi_value(fname, "database");
            fb_cgi_value(iname, "index");
            if (fname[0] != NULL && fname[0] != '/'){
               sprintf(word, "%s%s", path_dir, fname);
               strcpy(fname, word);
               }
            if (iname[0] != NULL && iname[0] != '/'){
               sprintf(word, "%s%s", path_dir, iname);
               strcpy(iname, word);
               }
            if (fname[0] != NULL){
	       cdb_db = fb_dballoc();
               fb_dbargs(fname, iname, cdb_db);
               withdatabase = 1;
               }

            strcpy(fname, path_trans);
            fb_cgi_foreach(mf_set_value);
            fb_cgi_checkbox(mf_set_checkbox);
            }
         else {
            /* normal (non cgi) path */
            if (withdatabase)
               fb_getargs(argc, argv, FB_ALLOCDB);
            else
               fb_getargs(argc, argv, FB_NODB);
            }

         cdb_lockmessage = 0;		/* turn off the lock() err message */
	 fb_allow_int();		/* allows instant piping */
/*
 * Before autoconf, this was _M_I386 from an SCO include file, I think.
 * Now its IGNORE_FPE set in configure.in for autoconf to use.
 */
#if IGNORE_FPE
         fb_ignore_fpe();
#endif
         /* set up standard pre constants used to signal */
         mf_set_constant("END", FB_END);
         mf_set_constant("ERROR", FB_ERROR);
         mf_set_constant("ABORT", FB_ABORT);
         mf_set_constant("NEXT", FB_ESIGNAL);
         mf_set_constant("PREV", FB_YSIGNAL);
         mf_set_constant("DEFAULT", FB_DEFAULT);
         mf_set_constant("NO_OP", FB_ERROR);
         mf_set_constant("AOK", 1);
         mf_set_constant("FB_AOK", FB_AOK);
         mf_set_constant("FB_ERROR", FB_ERROR);
         mf_set_constant("FB_END", FB_END);
         mf_set_constant("FB_ABORT", FB_ABORT);
         mf_set_constant("FB_NEXT", FB_ESIGNAL);
         mf_set_constant("FB_PREV", FB_YSIGNAL);
         mf_set_constant("FB_DEFAULT", FB_DEFAULT);
	 macro(fname, mscript);

         /* now do garbage collection */
         fb_gcollect_loadnode();
         fb_gcollect(n_ghead, c_ghead);
         fb_expunge_symtab(p_symtab);
         fb_expunge_symtab(g_symtab);
         fb_gcollect_m_pool();
         fb_gcollect_c_pool();
         fb_gcollect_s_pool();
	 fb_ender();
      }

/*
 *  macro - macro loop for a single dbase.
 */

   static macro(mname, mscript)
      char *mname, *mscript;
      
      {
	 fb_database *hp;
	 int i, tfd, st;

         if (initmacro(mname, mscript) == FB_ERROR)
            usage();
         hp = (fb_database *) NULL;
         if (withdatabase){
            if (cdb_limited){
               fb_serror(FB_MESSAGE, "Could not open database/index", NIL);
               return(FB_ERROR);
               }
            hp = cdb_db;
            tfd = open(cdb_db->dbase, 2);   /* test for write permissions */
            if (tfd > 0)
               mode = READWRITE;
            else
               mode = READ;
            close(tfd);
            st = fb_opendb(hp, mode, FB_ALLINDEX, FB_MAYBE_OPTIONAL_INDEX);
            if (st != FB_AOK){
               fb_serror(FB_MESSAGE, "Could not open database/index", NIL);
               return(FB_ERROR);
               }
            for (i = 0; i < hp->nfields; i++)
               fb_nounders(hp->kp[i]);
            }
         
         process(hp);

         /* no close on fd needed since always stdout */
         if (withdatabase)
	    fb_closedb(hp);
         return(FB_AOK);
      }

   static initmacro(mname, mscript)
      char *mname, *mscript;

      {
         int st;
         char buf[FB_MAXLINE];

         if (cdb_cgi_flag && cdb_usrlog > 5){
            sprintf(buf, "initmacro beginning - %s", mname);
            fb_usrlog_msg(buf);
            }
         if (mscript != NULL)
            st = fb_macroscript(mscript);
         else if (mname != NULL && mname[0] != NULL)
            st = fb_macrotree(mname);
         else
            st = fb_macrostdin();
         if (st == FB_AOK)
            e_winner = winner;
         if (cdb_cgi_flag && cdb_usrlog > 5){
            sprintf(buf, "initmacro ending - %s", mname);
            fb_usrlog_msg(buf);
            }
         return(st);
      }

/* 
 *  process - loop boot strap point using foreach ...
 */
 
   static process(hp)
      fb_database *hp;
      
      {
         int st = FB_ERROR;
	 
         n = 0L;

         /* if running with a database, locate and execute BEGIN section */
         if (withdatabase){
            s_exec = locate_section(S_BEGIN);
            if (s_exec != (fb_mnode *) NULL){
               mf_init_stack(main_sv);
               mf_make_frame(main_sv);
               m_verify_sub(s_exec);
               macro_statement(s_exec, main_sv);
               /* leave this sym table for other use */
               }
            }

         /* locate and execute BODY section */
         mf_init_stack(main_sv);
         mf_make_frame(main_sv);
         s_exec = locate_section(S_BODY);
         if (s_exec != (fb_mnode *) NULL){
            if (trace_flag)
               fb_macrotrace(s_exec, NULL);
            m_verify_sub(s_exec);
            if (withdatabase){
               if (hp->dindex == NULL || hp->dindex[0] == NULL || hp->ihfd < 0)
                  st = fb_foreach(hp, s_macro);
               else
                  st = fb_forxeach(hp, s_macro);
               }
            else
               s_macro((fb_database *) NULL);
            mf_destroy_frame(main_sv);
            m_destroy_call_lists(s_exec);
            }

         /* if running with a database, locate and execute END section */
         if (withdatabase){
            s_exec = locate_section(S_END);
            if (s_exec != (fb_mnode *) NULL){
               mf_init_stack(main_sv);
               mf_make_frame(main_sv);
               m_verify_sub(s_exec);
               macro_statement(s_exec, main_sv);
               }
            }

         return(st);
      }

/* 
 *  s_macro - called by foreach type mechanism
 */

   static s_macro(hp)
      fb_database *hp;
      
      {
	 n++;
         if (withdatabase)
            fb_set_autoindex(hp);
         if (s_exec != NULL){
            /*
             * if (cdb_cgi_flag && cdb_usrlog > 5)
             *   fb_usrlog_msg("process (7A)");
             */
            if (cdb_cgi_flag){
               if (setjmp(jmp_env) == 1){
                  /* value of 1 means return from longjmp */
                  alarm(0);
                  signal(SIGALRM, SIG_DFL);
                  fb_usrlog_msg("longjmp out of dbmacro execution");
                  return(FB_ERROR);
                  }
               signal(SIGALRM, dbmacro_sigalrm);
               alarm(cdb_locktime_cgi);
               }
            macro_statement(s_exec, main_sv);
            if (cdb_cgi_flag){
               alarm(0);
               signal(SIGALRM, SIG_DFL);
               }
            }
         if (main_sv->break_flag)
            fb_xerror(FB_MESSAGE, "Illegal `break'", NIL);
         if (main_sv->continue_flag)
            fb_xerror(FB_MESSAGE, "Illegal `continue'", NIL);
         if (cdb_write_it && mode == READWRITE && withdatabase){
            if (cdb_secure && !fb_record_permission(cdb_db, WRITE))
               cdb_write_it = 0;
            if (cdb_write_it && !read_only)
               writerec(hp);
            }
         if (main_sv->return_flag)
            return(FB_ERROR);
	 return(FB_AOK);
      }

/*
 * writerec - actual code to write/ceate the record.
 */

   static writerec(hp)
      fb_database *hp;

      {
         long rec;
         FILE *fs;

         rec = hp->rec;
         /* get lock of record first */
         if (fb_lock(rec, hp, FB_NOWAIT) != FB_AOK){ /* dont wait if locked */
            if (cdb_cgi_flag)
               fs = stdout;
            else
               fs = stderr;
            fprintf(fs, "Warning - record not written to disk: %ld\n",rec);
            return(0);
            }

         /* critical section - do not let this be interrupted */
	 signal(SIGINT, SIG_IGN);

	 fb_lock_head(hp);
	 fb_setdirty(hp, 1);
	 fb_allsync(hp);

	 /* if a putrec causes fb_xerror, rec 0 is still locked
	  * this will keep others out of the dbase until the
	  * problem can be fixed. this is a feature, not a bug.
	  */

	 /* fb_putlog(hp); */

	 if (fb_putrec(hp->rec, hp)==FB_ERROR)
	    fb_xerror(FB_WRITE_ERROR, SYSMSG[S_RECORD], hp->dbase);

	 /* fb_checklog(hp); */
	 fb_setdirty(hp, 0);
	 if (fb_put_autoindex(hp) == FB_ERROR)
	    fb_serror(FB_BAD_INDEX, hp->dbase, NIL);
	 fb_allsync(hp);
	 fb_unlock_head(hp);
	 fb_unlock(rec, hp);
	 cdb_write_it = 0;

         /* end critical section */
         signal(SIGINT, fb_onintr);
         return(1);
      }

/* 
 * usage message 
 */

   static void usage()
      {
         fb_xerror(FB_MESSAGE, USAGE, NIL);
      }

   static RETSIGTYPE dbmacro_sigalrm(disp)
      int disp;

      {
         (void) disp;
	 longjmp(jmp_env, 1);
      }
